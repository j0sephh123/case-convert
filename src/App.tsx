import { useMemo, useState, useCallback } from "react";
import { SetValue } from "./types";
import Textarea from "./components/Textarea/Textarea";
import * as transformers from "./transformers/transformers";

export default function App() {
  const [value, dispatchSetValue] = useState("");
  const setValue: SetValue = (newValue) => dispatchSetValue(newValue);

  const handleTransformToUpperCase = useCallback(() => {
    setValue((prevValue) => transformers.toUpperCase(prevValue));
  }, []);

  const handleTransformToLowerCase = useCallback(() => {
    setValue((prevValue) => transformers.toLowerCase(prevValue));
  }, []);

  const transformersList = useMemo(
    () => [
      {
        label: "UPPER CASE",
        onClick: handleTransformToUpperCase,
      },
      {
        onClick: handleTransformToLowerCase,
        label: "lower case",
      },
    ],
    [handleTransformToLowerCase, handleTransformToUpperCase]
  );

  return (
    <div className="p-8">
      <h1 className="text-3xl font-extrabold">Case Converter</h1>
      <div className="flex mt-10 justify-center">
        <Textarea value={value} onChange={setValue} />
      </div>
      <div className="flex justify-between mt-3 ml-4 mr-4">
        <div className="join gap-3">
          {transformersList.map(({ label, onClick }) => (
            <button
              key={label}
              onClick={onClick}
              className="btn btn-neutral normal-case"
            >
              {label}
            </button>
          ))}
        </div>
        <button onClick={() => setValue("")} className="btn">
          Reset
        </button>
      </div>
    </div>
  );
}
