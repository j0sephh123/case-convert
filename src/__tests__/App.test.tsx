import { render, screen, fireEvent } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import App from "../App";

describe("App", () => {
  it("updates the textarea value when typing", () => {
    const { asFragment } = render(<App />);
    const nativeTextarea = screen.getByRole("textbox");
    expect(nativeTextarea).toHaveValue("");

    fireEvent.change(nativeTextarea, {
      target: {
        value: "hello world",
      },
    });

    expect(nativeTextarea).toHaveValue("hello world");
    expect(asFragment()).toMatchSnapshot();
  });

  it("transforms the textarea value to upper case when 'UPPER CASE' button is clicked", () => {
    const { asFragment } = render(<App />);
    const nativeTextarea = screen.getByRole("textbox");
    const upperCaseButton = screen.getByText("UPPER CASE");

    fireEvent.change(nativeTextarea, {
      target: {
        value: "hello world",
      },
    });

    fireEvent.click(upperCaseButton);

    expect(nativeTextarea).toHaveValue("HELLO WORLD");
    expect(asFragment()).toMatchSnapshot();
  });

  it("transforms the textarea value to lower case when 'lower case' button is clicked", () => {
    const { asFragment } = render(<App />);
    const nativeTextarea = screen.getByRole("textbox");
    const lowerCaseButton = screen.getByText("lower case");

    fireEvent.change(nativeTextarea, {
      target: {
        value: "HELLO WORLD",
      },
    });

    fireEvent.click(lowerCaseButton);

    expect(nativeTextarea).toHaveValue("hello world");
    expect(asFragment()).toMatchSnapshot();
  });

  it("clears the textarea when 'Reset' button is clicked", () => {
    const { asFragment } = render(<App />);
    const nativeTextarea = screen.getByRole("textbox");
    const resetButton = screen.getByText("Reset");

    fireEvent.click(resetButton);

    expect(nativeTextarea).toHaveValue("");
    expect(asFragment()).toMatchSnapshot();
  });
});
