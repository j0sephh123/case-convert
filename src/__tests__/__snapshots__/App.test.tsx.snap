// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > clears the textarea when 'Reset' button is clicked 1`] = `
<DocumentFragment>
  <div
    class="p-8"
  >
    <h1
      class="text-3xl font-extrabold"
    >
      Case Converter
    </h1>
    <div
      class="flex mt-10 justify-center"
    >
      <textarea
        class="textarea textarea-lg w-full max-w-5xl border-zinc-400"
        rows="6"
      />
    </div>
    <div
      class="flex justify-between mt-3 ml-4 mr-4"
    >
      <div
        class="join gap-3"
      >
        <button
          class="btn btn-neutral normal-case"
        >
          UPPER CASE
        </button>
        <button
          class="btn btn-neutral normal-case"
        >
          lower case
        </button>
      </div>
      <button
        class="btn"
      >
        Reset
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App > transforms the textarea value to lower case when 'lower case' button is clicked 1`] = `
<DocumentFragment>
  <div
    class="p-8"
  >
    <h1
      class="text-3xl font-extrabold"
    >
      Case Converter
    </h1>
    <div
      class="flex mt-10 justify-center"
    >
      <textarea
        class="textarea textarea-lg w-full max-w-5xl border-zinc-400"
        rows="6"
      >
        hello world
      </textarea>
    </div>
    <div
      class="flex justify-between mt-3 ml-4 mr-4"
    >
      <div
        class="join gap-3"
      >
        <button
          class="btn btn-neutral normal-case"
        >
          UPPER CASE
        </button>
        <button
          class="btn btn-neutral normal-case"
        >
          lower case
        </button>
      </div>
      <button
        class="btn"
      >
        Reset
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App > transforms the textarea value to upper case when 'UPPER CASE' button is clicked 1`] = `
<DocumentFragment>
  <div
    class="p-8"
  >
    <h1
      class="text-3xl font-extrabold"
    >
      Case Converter
    </h1>
    <div
      class="flex mt-10 justify-center"
    >
      <textarea
        class="textarea textarea-lg w-full max-w-5xl border-zinc-400"
        rows="6"
      >
        HELLO WORLD
      </textarea>
    </div>
    <div
      class="flex justify-between mt-3 ml-4 mr-4"
    >
      <div
        class="join gap-3"
      >
        <button
          class="btn btn-neutral normal-case"
        >
          UPPER CASE
        </button>
        <button
          class="btn btn-neutral normal-case"
        >
          lower case
        </button>
      </div>
      <button
        class="btn"
      >
        Reset
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App > updates the textarea value when typing 1`] = `
<DocumentFragment>
  <div
    class="p-8"
  >
    <h1
      class="text-3xl font-extrabold"
    >
      Case Converter
    </h1>
    <div
      class="flex mt-10 justify-center"
    >
      <textarea
        class="textarea textarea-lg w-full max-w-5xl border-zinc-400"
        rows="6"
      >
        hello world
      </textarea>
    </div>
    <div
      class="flex justify-between mt-3 ml-4 mr-4"
    >
      <div
        class="join gap-3"
      >
        <button
          class="btn btn-neutral normal-case"
        >
          UPPER CASE
        </button>
        <button
          class="btn btn-neutral normal-case"
        >
          lower case
        </button>
      </div>
      <button
        class="btn"
      >
        Reset
      </button>
    </div>
  </div>
</DocumentFragment>
`;
